connectivity/service/src/main/java/org/eclipse/ditto/connectivity/service/mapping/CloudEventsMapper.java

```diff
@@ -0,0 +1,187 @@
+      final String data = payloadJson.getValue("data").get().toString();
+      return DittoJsonException.wrapJsonRuntimeException(
+          () -> ProtocolFactory.jsonifiableAdaptableFromJson(newObject(data)));
+    } else {
+      throw new Exception("Invalid CloudEvent");
+    }
+  }
+
+  public boolean checkHeaders(final ExternalMessage message) {
+    Map<String, String> headers = message.getHeaders();
+    if (headers.get(CE_ID) == null || headers.get(CE_SOURCE) == null
+        || headers.get(CE_TYPE) == null || headers.get(CE_SPECVERSION) == null) {
+      return false;
+    } else {
+      return true;
+    }
+  }
+
+
+  @Override
+  public List<Adaptable> map(final ExternalMessage message) {
+    // build Exception
+    MessageMappingFailedException mappingFailedException = MessageMappingFailedException.newBuilder(
+            message.findContentType().orElse("")).description("This is not a CloudEvent")
+        .dittoHeaders(DittoHeaders.of(message.getHeaders())).build();
+    if (message.findContentType().get().equals(contentType)) {
+      // extract message as String
+      final String payload = extractPayloadAsString(message);
+      try {
+        if (checkHeaders(message)) {
+          final JsonifiableAdaptable binaryAdaptable = DittoJsonException.wrapJsonRuntimeException(
+              () -> ProtocolFactory.jsonifiableAdaptableFromJson(newObject(payload)));
+          DittoHeaders headers = binaryAdaptable.getDittoHeaders();
+          return singletonList(
+              ProtocolFactory.newAdaptableBuilder(binaryAdaptable).withHeaders(headers).build());
+        }
+        if (validatePayload(payload)) {
+          final JsonifiableAdaptable adaptable = extractData(payload);
+          final DittoHeaders headers = adaptable.getDittoHeaders();
+          return singletonList(
+              ProtocolFactory.newAdaptableBuilder(adaptable).withHeaders(headers).build());
+        } else {
+          throw mappingFailedException;
+        }
+      } catch (Exception e) {
+        throw mappingFailedException;
+      }
+    }
+    return Collections.emptyList();
+  }
+
+  @Override
+  public DittoHeaders getAdditionalInboundHeaders(ExternalMessage message) {
+    return DittoHeaders.empty();
+  }
+
+  @Override
+  public List<ExternalMessage> map(final Adaptable adaptable) {
+    return List.of(ExternalMessageFactory.newExternalMessageBuilder(
+            Map.of(ExternalMessage.CONTENT_TYPE_HEADER, contentType))
+        .withTopicPath(adaptable.getTopicPath())
+        .withText(getExternalCloudEventSpecifications(adaptable))
+        .asResponse(isResponse(adaptable)).asError(isError(adaptable)).build());
+  }
+
+  private static String getExternalCloudEventSpecifications(Adaptable adaptable) {
+    final String OUTBOUNDID = UUID.randomUUID().toString();
+    JsonObject dataObject = JsonFactory.newObject(getJsonString(adaptable));
+    Map<JsonKey, JsonValue> outboundJson = new HashMap<>() {{
```

Comment:
please don't initialize Maps like this, this is an anti-pattern

> Because Double Brace Initialization (DBI) creates an anonymous class with a reference to the instance of the owning object, its use can lead to memory leaks if the anonymous inner class is returned and held by other objects. Even when there’s no leak, DBI is so obscure that it’s bound to confuse most maintainers.

Instead, just use a `JsonObjectBuilder` for building a new JsonObject:
```java
final JsonObject externalMessageObject = JsonObject.newBuilder()
            .set("data", dataObject)
            .set(...)
            .build();
```